### PR TITLE
feat(JA): Implementation for Job Attachments diff command

### PR DESCRIPTION
**Pull request by leongdl**
### What was the problem/requirement? (What/Why)
- Customers have requested APIs and CLI interfaces to use Job Attachments.
- We designed  new CLI + APIs:
   - manifest snapshot
   - manifest diff
   - manifest download
   - manifest upload
   - attachment upload
   - attachment download.

### What was the solution? (How)
- In this PR, we are adding the implementation for `manifest diff`

### What is the impact of this change?
- This new CLI + API will allow users to compare a Job Attachment manifest against the latest files in the local file system. Then customers can make a decision to make a new manifest file or not.

### How was this change tested?
- Manual testing
```
deadline manifest diff --root ./src --manifest ~/work/manifest/diff-cli-2024-10-08T16-40-40.manifest 
```
```
deadline manifest diff --root ./src --manifest /Users/leongdl/work/manifest/diff-cli-2024-10-08T16-40-40.manifest --json
```

- Have you run the unit tests?
    - Yes, `hatch run test` passes.
- Have you run the integration tests?
    - Yes `hatch run integ:test` passes.
```
================================================================================ 16 passed, 2 skipped, 4 warnings in 68.45s (0:01:08) ================================================================================
```

- Have you made changes to the `download` or `asset_sync` modules? If so, then it is highly recommended
  that you ensure that the docker-based unit tests pass.
- Not applicable, Diff only does manifest operations.

### Was this change documented?

- Are relevant docstrings in the code base updated?
    - Yes

- Has the README.md been updated? If you modified CLI arguments, for instance.
    - Readme and examples will be added once the feature stablizes.

### Does this PR introduce new dependencies?

This library is designed to be integrated into third-party applications that have bespoke and customized deployment environments. Adding dependencies will increase the chance of library version conflicts and incompatabilities. Please evaluate the addition of new dependencies. See the [Dependencies](https://github.com/aws-deadline/deadline-cloud/blob/mainline/DEVELOPMENT.md#dependencies) section of DEVELOPMENT.md for more details.

*   [ ] This PR adds one or more new dependency Python packages. I acknowledge I have reviewed the considerations for adding dependencies in [DEVELOPMENT.md](https://github.com/aws-deadline/deadline-cloud/blob/mainline/DEVELOPMENT.md#dependencies).
*   [X] This PR does not add any new dependencies.

### Is this a breaking change?
- No.

### Does this change impact security?
- Yes, this will be security reviewed.

----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

### CODE REVIEW

File: src/deadline/job_attachments/api/manifest.py [hunk 1]
```diff
@@ -29,6 +36,37 @@
 """
 
 
+def _glob_files(
```
Comment (leongdl) [inline src/deadline/job_attachments/api/manifest.py:38]: Refactored away to share code.

File: test/unit/deadline_client/cli/test_cli_manifest_diff.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+from click.testing import CliRunner
+from deadline.client.cli._groups.manifest_group import cli_manifest
+import pytest
+import tempfile
+
+from deadline.client.cli import main
+
+
+@pytest.mark.skip("Random Failure with no credentials on Github")
+class TestSnapshot:
```
Comment (leongdl) [inline test/unit/deadline_client/cli/test_cli_manifest_diff.py:18]: There is something odd with calling CLI commands in unit test. As integration tests they pass but I am not sure why. I'm going to comment this out for now and then once submissions open I may move this over as an integ test.

File: src/deadline/client/cli/_groups/manifest_group.py [hunk 3]
```diff
@@ -139,6 +142,12 @@ def manifest_snapshot(
     default=None,
     help="Include and exclude config of files and directories to include and exclude. Can be a json file or json string.",
 )
+@click.option(
+    "--force-rehash",
+    default=False,
+    is_flag=True,
+    help="Rehash all files to compare using file hashes.",
+)
```
Comment (godobyte) [inline src/deadline/client/cli/_groups/manifest_group.py:151]: I don't see this being added in for snapshot command, do you mind adding it before the next release?

File: src/deadline/client/cli/_groups/manifest_group.py [hunk 4]
```diff
@@ -147,13 +156,43 @@ def manifest_diff(
...
 ):
     """
     Check file differences between a directory and specified manifest.
     """
-    raise NotImplementedError("This CLI is being implemented.")
+    logger: ClickLogger = ClickLogger(is_json=json)
+    if not os.path.isfile(manifest):
+        raise NonValidInputError(f"Specified manifest file {manifest} does not exist. ")
+
+    if not os.path.isdir(root):
```
Comment (godobyte) [inline src/deadline/client/cli/_groups/manifest_group.py:171]: nit - we should be consistent with option naming `root` or `root-dir` for attachment and manifest commands.

File: src/deadline/client/cli/_groups/manifest_group.py [hunk 5]
```diff
@@ -147,13 +156,43 @@ def manifest_diff(
...
+        raise NonValidInputError(f"Specified root directory {root} does not exist. ")
+
+    # Perform the diff.
+    differences = _manifest_diff(
```
Comment (godobyte) [inline src/deadline/client/cli/_groups/manifest_group.py]: nit - no typing for `differences`.

File: src/deadline/client/cli/_groups/manifest_group.py [hunk 6]
```diff
@@ -147,13 +156,43 @@ def manifest_diff(
...
+        include=include,
+        exclude=exclude,
+        include_exclude_config=include_exclude_config,
+        force_rehash=force_rehash,
+        logger=logger,
+    )
+
+    # Print results to console.
+    if json:
+        logger.json(dataclasses.asdict(differences), indent=4)
```
Comment (godobyte) [inline src/deadline/client/cli/_groups/manifest_group.py:187]: nit - `json` function already does json check?

File: src/deadline/client/cli/_groups/manifest_group.py [hunk 7]
```diff
@@ -147,13 +156,43 @@ def manifest_diff(
...
+    else:
+        logger.echo(f"\n{root}")
```
Comment (godobyte) [inline src/deadline/client/cli/_groups/manifest_group.py]: nit - could we add more description for this log?

File: src/deadline/job_attachments/api/manifest.py (hunk 1 above)
Comment (godobyte) [reply src/deadline/job_attachments/api/manifest.py:38]: Could this be in `_glob.py`?

File: src/deadline/job_attachments/_diff.py [hunk 8]
```diff
@@ -117,47 +118,184 @@ def compare_manifest(
...
+    return_root_relative_path: bool = True,
+) -> List[Tuple[str, FileStatus]]:
     """
     Perform a fast difference of the current list of files to a previous manifest to diff against using time stamps and file sizes.
     :param root: Root folder of files to diff against.
     :param current_files: List of files to compare with.
     :param diff_manifest: Manifest containing files to diff against.
+    :param return_root_relative_path: File Path to return, either relative to root or full.
+    :param logger: logger.
     :return List[str]: List of files that are new, or modified.
```
Comment (godobyte) [inline src/deadline/job_attachments/_diff.py]: nit - doc need to be updated too.

File: src/deadline/job_attachments/_diff.py [hunk 9]
```diff
@@ -117,47 +118,184 @@ def compare_manifest(
...
     """
-    changed_paths: List[str] = []
+
+    def select_path(full_path: str, relative_path: str, return_root_relative_path: bool):
```
Comment (godobyte) [inline src/deadline/job_attachments/_diff.py:138]: Q - what's the use case for full or relative path?

File: src/deadline/job_attachments/_diff.py [hunk 10]
```diff
@@ -117,47 +118,184 @@ def compare_manifest(
...
+        "DIRECTORY": "\033[80m",  # grey
+    }
+
+    # Tooltips:
+    TOOLTIPS = {
+        FileStatus.NEW: " +",  # added files
+        FileStatus.DELETED: " -",  # deleted files
+        FileStatus.MODIFIED: " M",  # modified files
+        FileStatus.UNCHANGED: "",  # unchanged files
+    }
```
Comment (godobyte) [inline src/deadline/job_attachments/_diff.py:229]: Could we model file status related configurations better?

Some idea - https://stackoverflow.com/questions/59916345/adding-a-property-to-an-enum

File: src/deadline/job_attachments/_diff.py [hunk 11]
```diff
@@ -117,47 +118,184 @@ def compare_manifest(
...
+            relative_path = str(Path(file).relative_to(root))
+            add_to_tree(
+                relative_path,
+                get_file_status(relative_path, manifest_diff),
+            )
+        return directory_tree
+
+    directory_tree = build_directory_tree(all_files)
+    print_tree(directory_tree)
+    logger.info("")
```
Comment (godobyte) [inline src/deadline/job_attachments/_diff.py:302]: Is this on purpose?

File: src/deadline/job_attachments/api/manifest.py [hunk 12]
```diff
@@ -128,3 +173,78 @@ def _manifest_snapshot(
...
+    input_paths = [Path(p) for p in input_files]
+
+    # Placeholder Asset Manager
+    asset_manager = S3AssetManager()
+
+    # parse the given manifest to compare against.
+    local_manifest_object: BaseAssetManifest
+    with open(manifest) as input_file:
+        manifest_data_str = input_file.read()
+        local_manifest_object = decode_manifest(manifest_data_str)
```
Comment (godobyte) [inline src/deadline/job_attachments/api/manifest.py:212]: Should we add file check and error handling here?

File: src/deadline/job_attachments/api/manifest.py [hunk 13]
```diff
@@ -128,3 +173,78 @@ def _manifest_snapshot(
...
+        fast_diff: List[Tuple[str, FileStatus]] = _fast_file_list_to_manifest_diff(
+            root=root, current_files=input_files, diff_manifest=local_manifest_object, logger=logger
+        )
+        for fast_diff_item in fast_diff:
+            if fast_diff_item[1] == FileStatus.MODIFIED:
+                output.modified.append(fast_diff_item[0])
+            elif fast_diff_item[1] == FileStatus.NEW:
+                output.new.append(fast_diff_item[0])
+            elif fast_diff_item[1] == FileStatus.DELETED:
+                output.deleted.append(fast_diff_item[0])
```
Comment (godobyte) [inline src/deadline/job_attachments/api/manifest.py:248]: The comparison for both cases look very similar, maybe we could refactor or use helper?

File: test/unit/deadline_client/cli/test_cli_manifest_diff.py (hunk 2 above)
Comment (godobyte) [reply test/unit/deadline_client/cli/test_cli_manifest_diff.py:18]: These look like integ tests to me. I'd imagine cli test tests the wrapper and mock call to the api (such as `_manifest_diff`).

File: src/deadline/client/cli/_groups/manifest_group.py (hunk 3 above)
Comment (leongdl) [reply src/deadline/client/cli/_groups/manifest_group.py:151]: Yes - I will, this PR is specifically for diff. My Next PR will also add it to snapshot for equivalence.

File: src/deadline/client/cli/_groups/manifest_group.py (hunk 4 above)
Comment (leongdl) [reply src/deadline/client/cli/_groups/manifest_group.py:171]: I agree! Lets do a quick pass after we merge in to catch these inconsistencies.

Root or root dir?

File: src/deadline/client/cli/_groups/manifest_group.py (hunk 5 above)
Comment (leongdl) [reply src/deadline/client/cli/_groups/manifest_group.py]: Will fix.

File: src/deadline/client/cli/_groups/manifest_group.py (hunk 6 above)
Comment (leongdl) [reply src/deadline/client/cli/_groups/manifest_group.py:187]: What do you mean? logger.json takes in a dictionary so it prints out JSON.

File: src/deadline/client/cli/_groups/manifest_group.py (hunk 7 above)
Comment (leongdl) [reply src/deadline/client/cli/_groups/manifest_group.py]: Sure.

File: src/deadline/client/cli/_groups/manifest_group.py (hunk 4 above)
Comment (godobyte) [reply src/deadline/client/cli/_groups/manifest_group.py:171]: I'd suggest root. Let's do a pass with stakeholders for the naming, and update them altogether.

File: src/deadline/client/cli/_groups/manifest_group.py (hunk 6 above)
Comment (godobyte) [reply src/deadline/client/cli/_groups/manifest_group.py:187]: The `json` function in clicklogger already check `if json`, this line doesn't need to be under the json check again.

File: src/deadline/client/cli/_groups/manifest_group.py (hunk 4 above)
Comment (leongdl) [reply src/deadline/client/cli/_groups/manifest_group.py:171]: We agreed with `root` (this PR's name) and also will do a pass potentially to cleanup.

File: src/deadline/job_attachments/_diff.py (hunk 8 above)
Comment (leongdl) [reply src/deadline/job_attachments/_diff.py]: updated! Its a

File: src/deadline/job_attachments/_diff.py (hunk 9 above)
Comment (leongdl) [reply src/deadline/job_attachments/_diff.py:138]: in the snapshot case, it requires the full path. Snapshot (and related code) uses the full path. Diff on the otherhand uses relative path. I didn't want to duplicate the code, so I used a function toggle to allow returning in either format easily.

File: src/deadline/job_attachments/api/manifest.py (hunk 1 above)
Comment (leongdl) [reply src/deadline/job_attachments/api/manifest.py:38]: No - because this function is local only to the CLIs and only used by the CLIs, so I am not moving this to a shared code. It is shared code to process the CLI input. 

The shared function or API is `_process_glob_inputs` below.

File: test/unit/deadline_client/cli/test_cli_manifest_diff.py (hunk 2 above)
Comment (leongdl) [reply test/unit/deadline_client/cli/test_cli_manifest_diff.py:18]: Yes, somehow there is something odd with them being unit tests. I will move them to be integ tests once the code is merged. I didn't want to merge in some functionality with potentially flaky integ tests.  (I'll follow up after this PR)

File: src/deadline/job_attachments/_diff.py (hunk 11 above)
Comment (leongdl) [reply src/deadline/job_attachments/_diff.py:302]: Yes - this is part of the "pretty print" to print an empty line.

File: src/deadline/job_attachments/api/manifest.py (hunk 13 above)
Comment (leongdl) [reply src/deadline/job_attachments/api/manifest.py:248]: Yes and no, the loop is the same, but the data structure it processes is the different. So I dont' see much refactoring options here.

File: src/deadline/job_attachments/_diff.py [hunk 14]
```diff
@@ -117,47 +118,185 @@ def compare_manifest(
...
+    :return List[Tuple[str, FileStatus]]: List of Tuple containing the file path and FileStatus pair.
     """
-    changed_paths: List[str] = []
+
+    # Select either relative or absolut path for results.
+    def select_path(full_path: str, relative_path: str, return_root_relative_path: bool):
+        if return_root_relative_path:
+            return relative_path
+        else:
+            return full_path
```
Comment (benl-2023) [inline src/deadline/job_attachments/_diff.py:142]: nit: could be 1 liner with ternary operator

**Review by benl-2023 — APPROVED** [review]

File: src/deadline/job_attachments/api/manifest.py (hunk 13 above)
Comment (godobyte) [reply src/deadline/job_attachments/api/manifest.py:248]: Maybe something like?
 
```
def process_output(status, path, output):
    if status == FileStatus.MODIFIED:
        output.modified.append(path)
    elif status == FileStatus.NEW:
        output.new.append(path)
    elif status == FileStatus.DELETED:
        output.deleted.append(path)
    else:
        raise InvalidStatusError()
```

File: src/hello
Comment (godobyte) [inline src/hello:1]: Is this file needed?

**Review by godobyte — APPROVED** [review]

File: src/deadline/client/cli/_groups/manifest_group.py (hunk 6 above)
Comment (leongdl) [reply src/deadline/client/cli/_groups/manifest_group.py:187]: Oh! Thats what you meant. Yeah I recognize that in this case, but given there's a line for the pretty print option it was easier to already early escape it here. 

The pretty print does alot more un-necessary work so I wanted to avoid running that code for no reason.

File: src/deadline/job_attachments/_diff.py (hunk 14 above)
Comment (leongdl) [reply src/deadline/job_attachments/_diff.py:142]: True :) Very pythonic.

File: src/deadline/job_attachments/_diff.py (hunk 10 above)
Comment (leongdl) [reply src/deadline/job_attachments/_diff.py:229]: I agree- let me look into this refactor after we complete all the CLIs and name cleanups.

File: src/deadline/job_attachments/api/manifest.py (hunk 13 above)
Comment (leongdl) [reply src/deadline/job_attachments/api/manifest.py:248]: Sure, I can do that cleanup but I wanted to avoid nested functions :(

File: src/deadline/job_attachments/api/manifest.py (hunk 12 above)
Comment (leongdl) [reply src/deadline/job_attachments/api/manifest.py:212]: `decode_manifest` raises errors, so if this failed, it will just fail the program outright. There's no need to do more error checks redundant.
